pkgs/build-support/buildenv/default.nix

```diff
@@ -13,114 +13,142 @@ let
+        # directories in the list is not symlinked.
+        pathsToLink ? [ "/" ],
+
+        # The package outputs to include. By default, only the default
+        # output is included.
+        extraOutputsToInstall ? [ ],
+
+        # Root the result in directory "$out${extraPrefix}", e.g. "/share".
+        extraPrefix ? "",
+
+        # Shell commands to run after building the symlink tree.
+        postBuild ? "",
+
+        # Additional stdenv.mkDerivation arguments
+        # such as nativeBuildInputs/buildInputs for postBuild dependencies.
+        derivationArgs ? { },
+
+        # Placeholder name arguments.
+        name ? null,
+        pname ? null,
+        version ? null,
+
+        # `stdenv.mkDerivation` args before introducing derivationArgs.
+        nativeBuildInputs ? null,
+        buildInputs ? null,
+        passthru ? null,
+        meta ? null,
+      }@args:
+      let
+        compatArgs = getOptionalAttrs [
+          "nativeBuildInputs"
+          "buildInputs"
+          "passthru"
+          "meta"
+        ] args;
+
+        chosenOutputs = map (drv: {
+          paths =
+            # First add the usual output(s): respect if user has chosen explicitly,
+            # and otherwise use `meta.outputsToInstall`. The attribute is guaranteed
+            # to exist in mkDerivation-created cases. The other cases (e.g. runCommand)
+            # aren't expected to have multiple outputs.
+            (
+              if
+                (!drv ? outputSpecified || !drv.outputSpecified) && drv.meta.outputsToInstall or null != null
+              then
+                map (outName: drv.${outName}) drv.meta.outputsToInstall
+              else
+                [ drv ]
+            )
+            # Add any extra outputs specified by the caller of `buildEnv`.
+            ++ lib.filter (p: p != null) (
+              builtins.map (outName: drv.${outName} or null) finalAttrs.extraOutputsToInstall
+            );
+          priority = drv.meta.priority or lib.meta.defaultPriority;
+        }) finalAttrs.paths;
+
+        pathsForClosure = lib.pipe chosenOutputs [
+          (map (p: p.paths))
+          lib.flatten
+          (lib.remove null)
+        ];
+      in
+      derivationArgs
+      // compatArgs
+      // {
         inherit
+          extraOutputsToInstall
           manifest
           ignoreCollisions
           checkCollisionContents
           ignoreSingleFileOutputs
-          passthru
-          meta
+          includeClosures
+          paths
           pathsToLink
           extraPrefix
           postBuild
-          nativeBuildInputs
-          buildInputs
           ;
         pkgs = builtins.toJSON chosenOutputs;
-        extraPathsFrom = lib.optional includeClosures (writeClosure pathsForClosure);
-        preferLocalBuild = true;
-        allowSubstitutes = false;
+        extraPathsFrom = lib.optional finalAttrs.includeClosures (writeClosure pathsForClosure);
```

Comment:
I noticed that `extraPathsFrom` is treated as a single file in `buildenv/builder.pl`. We should correct it and make it a string instead of a list.

Reply:
Applied.